shared/middleware/files_api.rb

```diff
@@ -498,14 +504,16 @@ def files_put_file(encrypted_channel_id, filename, body)
       manifest_is_unchanged = false
     end
 
-    # if we're also deleting a file (on rename), remove it from the manifest
-    if params['delete']
+    # Handle any deletion request, unless a case-only rename is occuring.
+    if params['delete'] && params['delete'].downcase != downcased_filename
       manifest_delete_comparison_filename = CGI.unescape(params['delete']).downcase
       reject_result = manifest.reject! {|e| e['filename'].downcase == manifest_delete_comparison_filename}
       manifest_is_unchanged = false unless reject_result.nil?
+
+      bucket.delete(encrypted_channel_id, params['delete'].downcase) if params['delete']
```

Comment:
I believe the original intent is to avoid letting the manifest ever point to a missing file, in which case the strange ordering makes sense.  We're optimizing for reliability of the manifest. Think of it this way:

* When adding a file, you want to update the manifest last.
  ```
  put_file
  update_manifest
  ```
  - If an error occurs during `put_file` nothing happens and the world is in a good state.
  - If an error occurs during `update_manifest` we have an orphaned file, but at least the manifest is still in a good state.
* When deleting a file you want to update the manifest first.
  ```
  update_manifest
  delete_file
  ```
  - If an error occurs during `update_manifest` nothing happens and the world is in a good state
  - If an error occurs during `delete_file` we have an orphaned file, but at least the manifest is in a good state.
* When moving a file, we update the manifest in the middle because of the combination of the above rules.
  ```
  put_file
  update_manifest
  delete_file
  ```
  - If an error occurs in `put_file` nothing happens and the world is in a good state.
  - If an error occurs in `update_manifest` we've orphaned the new file but the manifest and the old file are fine.
  - If an error occurs in `delete_file` we've orphaned the old file but the manifest and the new file are fine.

With your new setup there's a possible bad state:
```
put_file
delete_file
update_manifest
```
If the error occurs during `update_manifest` we've got a manifest pointing to a deleted file, which results in user bad.  Orphaned files don't result in user bad.